cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -324,25 +324,245 @@ module ProductFlow {
+        or
+        exists(Flow1::PathNode mid |
+          revRevLocalPathStep1(mid) and
+          localPathStep1(n, mid)
+        )
+      )
+    }
+
+    pragma[assume_small_delta]
+    private predicate revLocalPathStep1(Flow1::PathNode n1, Flow1::PathNode n2) {
+      revRevLocalPathStep1(n1) and
+      revRevLocalPathStep1(n2) and
+      localPathStep1(n1, n2)
+    }
+
+    pragma[nomagic]
+    private predicate revFwdLocalPathStep2(Flow2::PathNode n) {
+      fwdReachableInterprocEntry(_, n) and
+      (
+        interprocEdgePair(_, _, _, n)
+        or
+        isSourcePair(_, n)
+      )
       or
-      exists(
-        Flow1::PathNode midEntry1, Flow2::PathNode midEntry2, Flow1::PathNode midExit1,
-        Flow2::PathNode midExit2
+      exists(Flow2::PathNode mid |
+        revFwdLocalPathStep2(mid) and
+        localPathStep2(mid, n)
+      )
+    }
+
+    pragma[nomagic]
+    private predicate revRevLocalPathStep2(Flow2::PathNode n) {
+      revFwdLocalPathStep2(n) and
+      (
+        revReachableInterprocEntry(_, n)
+        or
+        exists(Flow2::PathNode mid |
+          revRevLocalPathStep2(mid) and
+          localPathStep2(n, mid)
+        )
+      )
+    }
+
+    pragma[assume_small_delta]
+    private predicate revLocalPathStep2(Flow2::PathNode n1, Flow2::PathNode n2) {
+      revRevLocalPathStep2(n1) and
+      revRevLocalPathStep2(n2) and
+      localPathStep2(n1, n2)
+    }
+
+    private predicate revLocalPathStep1SuccPlus(Flow1::PathNode n1, Flow1::PathNode n2) =
+      fastTC(revLocalPathStep1/2)(n1, n2)
+
+    private predicate revLocalPathStep2SuccPlus(Flow2::PathNode n1, Flow2::PathNode n2) =
+      fastTC(revLocalPathStep2/2)(n1, n2)
+
+    private predicate revLocalPathStep1Tc(Flow1::PathNode n1, Flow1::PathNode n2) {
+      revLocalPathStep1SuccPlus(n1, n2) or n1 = n2
+    }
+
+    private predicate revLocalPathStep2Tc(Flow2::PathNode n1, Flow2::PathNode n2) {
+      revLocalPathStep2SuccPlus(n1, n2) or n1 = n2
+    }
+
+    pragma[assume_small_delta]
+    private predicate revIsSuccessor(
+      Flow1::PathNode pred1, Flow2::PathNode pred2, Flow1::PathNode succ1, Flow2::PathNode succ2
+    ) {
+      revReachableInterprocEntry(succ1, succ2) and
+      exists(Flow1::PathNode mid1, Flow2::PathNode mid2 |
+        fwdReachableInterprocEntry(mid1, mid2) and
+        revLocalPathStep1Tc(mid1, succ1) and
+        revLocalPathStep2Tc(mid2, succ2)
       |
-        reachableInterprocEntry(source1, source2, midEntry1, midEntry2) and
-        interprocEdgePair(midExit1, midExit2, node1, node2) and
-        localPathStep1*(midEntry1, midExit1) and
-        localPathStep2*(midEntry2, midExit2)
+        interprocEdgePair(pred1, pred2, mid1, mid2)
+        or
+        isSourcePair(mid1, mid2) and
+        pred1 = mid1 and
+        pred2 = mid2
       )
     }
 
+    pragma[assume_small_delta]
+    pragma[nomagic]
+    private predicate revReachableInterprocEntry(Flow1::PathNode node1, Flow2::PathNode node2) {
```

Comment:
You are making the reverse calculation much more complicated than it has to be. Instead of mirroring the forward calculation, you can make things much simpler by using `fwdIsSuccessor` as your step relation. That way there will be no need to do anything with `revLocalPathStep*`. You'll get slightly different node-pairs, but that's because the code as-is doesn't calculate `revReachableInterprocEntry` as advertised but rather the exits, so your `revReachableInterprocEntry` confusingly isn't a subset of `fwdReachableInterprocEntry`. With this simplification the reverse flow will simply be a subset of `fwdReachableInterprocEntry` and the node pairs from the forward and reverse calculations will be consistent.

Reply:
Thanks for the comments. I've force-pushed a version that hopefully implements the simplifications you had in mind 🤞.